File: core/src/main/kotlin/com/piotrwalkusz/smartlaw/core/model/element/common/type/InterfaceRefTemplate.kt
package com.piotrwalkusz.smartlaw.core.model.element.common.type

import com.piotrwalkusz.smartlaw.core.model.common.Id
import com.piotrwalkusz.smartlaw.core.model.template.Template

data class InterfaceRefTemplate(
        val interfaceId: Template<Id>,
        val parameters: Template<List<TypeTemplate>>
) : TypeTemplate

File: frontend/src/model/InterfaceRefTemplate.ts
import Type, { TypeKind } from "./Type";
import Id, { decodeId } from "./Id";
import { List } from "immutable";
import { decodeList } from "../utils/Decoders";
import Template, { decodeTemplate } from "./Template";
import TypeTemplate, { decodeTypeTemplate } from "./TypeTemplate";

export const decodeInterfaceRef = (json: any): InterfaceRefTemplate => {
  return {
    type: TypeKind.InterfaceRef,
    interfaceId: decodeTemplate(json.interfaceId, decodeId),
    parameters: decodeTemplate(json.parameters, (json) => decodeList(json, decodeTypeTemplate)),
  };
};

export default interface InterfaceRefTemplate extends Type {
  interfaceId: Template<Id>;
  parameters: Template<List<TypeTemplate>>;
}
